fuzz/fuzz.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (C) 2022 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include "ass.h"
+#include "ass_types.h"
+
+#define FUZZMODE_STANDALONE       0
+#define FUZZMODE_AFLXX_SHAREDMEM  1
+#define FUZZMODE_LIBFUZZER        2
+#ifndef ASS_FUZZMODE
+    #define ASS_FUZZMODE FUZZMODE_STANDALONE
+#endif
+
+ASS_Library *ass_library;
+ASS_Renderer *ass_renderer;
+
+void msg_callback(int level, const char *fmt, va_list va, void *data)
+{
+#if ASS_FUZZMODE == FUZZMODE_STANDALONE
+    if (level > 6) return;
+    printf("libass: ");
+    vprintf(fmt, va);
+    printf("\n");
+#endif
+}
+
+static bool init(void)
+{
+    ass_library = ass_library_init();
+    if (!ass_library) {
+        printf("ass_library_init failed!\n");
+        return false;
+    }
+
+    ass_set_message_cb(ass_library, msg_callback, NULL);
+
+    ass_renderer = ass_renderer_init(ass_library);
+    if (!ass_renderer) {
+        ass_library_done(ass_library);
+        ass_library = NULL;
+        printf("ass_renderer_init failed!\n");
+        return false;
+    }
+
+    ass_set_fonts(ass_renderer, NULL, "sans-serif",
+                  ASS_FONTPROVIDER_AUTODETECT, NULL, 1);
+    ass_set_frame_size(ass_renderer, 854, 480);
```

Comment:
I think it's better to make some variations here to check bitmap alignment issues and such. Maybe we can use some seed value from the start of the input buffer (or command line parameter for standalone mode) to generate random sizes in some range.

Reply:
Other than possibly very small and very large frame/storage_sizes triggering something about the (re)alloc strategy, it’s not clear too me what code paths could be unlocked by varying the size not reachable with this fixed value. Also varying the sizes and codepaths based on some random(ly selected) bytes might confuse the fuzzing engines code coverage based mutator, doing more harm than good.

Reply:
I mean not real random, but `deterministic_noise_function(seed)`, so it should be perfectly ordinary for fuzzing engine to mutate `seed`.

Reply:
I was already assuming deterministic behaviour for identical seed. AFL and libFuzzer aren't just doing naïve mutations, though they try to maximize coverage. If now alteration of say the last few bits changes the initial setup and thus possibly a lot of branches, this might interfere with their ability to explore the new paths deeper.
Furthermore, rendering for larger frame sizes is more expensive and as noted before, I don't see much benefit in it other than for extremly small or extremly large frame_sizes. All other effects of different frame sizes should also be able to be triggered by altering the input.

*(Side note: as for your other suggestions, I want to measure the performance impact first to decide if this will be always be enabled or conditionally via another macro define. It might take a few days until I get to that.)*